src/goto-instrument/code_contracts.cpp

```diff
@@ -188,6 +206,29 @@ void code_contractst::check_apply_invariants(
     converter.goto_convert(assumption, havoc_code, mode);
   }
 
+  // Temporary variables that store the variant before and after the loop
+  symbol_exprt old_variant_symbol =
+    new_tmp_symbol(
+      variant.type(), loop_head->source_location, function_name, mode)
+      .symbol_expr();
+  symbol_exprt new_variant_symbol =
+    new_tmp_symbol(
+      variant.type(), loop_head->source_location, function_name, mode)
+      .symbol_expr();
+
+  if(!variant.is_nil())
+  {
+    // Generate: a declaration of the temporary variable that stores the loop
+    // variant before the loop entry
+    havoc_code.add(
+      goto_programt::make_decl(old_variant_symbol, loop_head->source_location));
+
+    // Generate: an assignment to store the loop variant before the loop entry
+    code_assignt old_variant_assignment{old_variant_symbol, variant};
+    old_variant_assignment.add_source_location() = loop_head->source_location;
+    converter.goto_convert(old_variant_assignment, havoc_code, mode);
```

Comment:
@tautschnig is this use of `goto_convert` fitting with how and when / where it is supposed to be used?